src/base/net/reverseresolution.cpp

```diff
@@ -52,37 +52,36 @@ ReverseResolution::ReverseResolution(QObject *parent)
 
 ReverseResolution::~ReverseResolution()
 {
-    qDebug("Deleting host name resolver...");
+    // abort on-going lookups instead of waiting them
+    for (auto iter = m_lookups.cbegin(); iter != m_lookups.cend(); ++iter)
+        QHostInfo::abortHostLookup(iter.key());
 }
 
-void ReverseResolution::resolve(const QString &ip)
+void ReverseResolution::resolve(const QHostAddress &ip)
 {
-    if (m_cache.contains(ip)) {
-        const QString &hostname = *m_cache.object(ip);
-        qDebug() << "Resolved host name using cache: " << ip << " -> " << hostname;
-        if (isUsefulHostName(hostname, ip))
-            emit ipResolved(ip, hostname);
-    }
-    else {
-        // Actually resolve the ip
-        m_lookups.insert(QHostInfo::lookupHost(ip, this, &ReverseResolution::hostResolved), ip);
+    const QString *hostname = m_cache.object(ip);
+    if (hostname) {
+        emit ipResolved(ip, *hostname);
+        return;
     }
+
+    // do reverse lookup: IP -> hostname
+    const int lookupId = QHostInfo::lookupHost(ip.toString(), this, &ReverseResolution::hostResolved);
```

Comment:
What if there are multiple simultaneous requests for the same IP (sequential calls to this function with same IP)?

Reply:
I cannot really know by just reading the Qt docs... But I guess a sane library would avoid the repetitive queries.
https://github.com/qt/qtbase/blob/1e654450157ca422833d48c0faf4fe0836c3125f/src/network/kernel/qhostinfo.cpp
https://github.com/qt/qtbase/blob/1e654450157ca422833d48c0faf4fe0836c3125f/src/network/kernel/qhostinfo_p.h

Reply:
>But I guess a sane library would avoid the repetitive queries.

Can we only hope for library behavior that is not explicitly declared, or should we prevent it ourselves at the application level?

Reply:
I have no strong opinion. And I'm not sure that is really a big problem in qbt...

Reply:
This is not a problem for me personally. But I can imagine that in some environment there may be many peers using a single address, but different ports. Then qBittorrent (possible) will send N "host resolution" requests and update all N peer items N times...
Although it is simple enough to prevent simultaneous requests to the same address, I will not insist on this (maybe Qt really does it).

Reply:
>update all N peer items N times...

This problem seems to exist on its own (even if simultaneous requests are not actually executed, or the hostname is taken from the cache).